File: observability-study/sleuth/consumer/src/main/java/com/bso/consumer/infra/aws/sender/SqsMessageSender.java
package com.bso.consumer.infra.aws.sender;

import brave.Tracer;
import brave.Tracing;
import brave.propagation.Propagation;
import brave.propagation.TraceContext;
import io.awspring.cloud.messaging.core.QueueMessagingTemplate;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;

@Component
public class SqsMessageSender implements MessageSender {
    private final QueueMessagingTemplate queueMessagingTemplate;
    private final Tracer tracer;
    private final TraceContext.Injector<Map<String, Object>> injector;

    public SqsMessageSender(Tracing tracing, QueueMessagingTemplate queueMessagingTemplate, Tracer tracer) {
        this.queueMessagingTemplate = queueMessagingTemplate;
        this.tracer = tracer;
        Propagation.Setter<Map<String, Object>, String> setter = (request, key, value) -> request.put(key, value);
        this.injector = tracing.propagation().injector(setter);
    }

    @Override
    public <T> void send(T message, String queue) {
        Map<String, Object> headers = new HashMap<>();
        injector.inject(tracer.currentSpan().context(), headers);
        queueMessagingTemplate.convertAndSend(queue, message, headers);
    }
}

File: observability-study/open-telemetry/consumer/src/main/kotlin/com/bso/opentelemetry/consumer/infra/aws/sqs/SqsMessageSender.kt
package com.bso.opentelemetry.consumer.infra.aws.sqs

import com.bso.opentelemetry.consumer.infra.monitoring.sqs.SqsTracingInjector
import com.fasterxml.jackson.databind.ObjectMapper
import io.opentelemetry.api.OpenTelemetry
import io.opentelemetry.context.Context
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import software.amazon.awssdk.services.sqs.SqsAsyncClient
import software.amazon.awssdk.services.sqs.model.GetQueueUrlRequest
import software.amazon.awssdk.services.sqs.model.MessageAttributeValue
import software.amazon.awssdk.services.sqs.model.SendMessageRequest

@Component
class SqsMessageSender(
    private val sqsAsyncClient: SqsAsyncClient,
    private val objectMapper: ObjectMapper,
    private val openTelemetry: OpenTelemetry
) {
    private val logger: Logger by lazy { LoggerFactory.getLogger(this::class.java) }
    private val cacheLock = Any()
    private val cache: Cache<String, String> = Cache()

    fun send(queueName: String, message: Any) {
        val queueUrl: String = getQueueFromCacheOrSdk(queueName)
        val request = SendMessageRequest.builder()
            .queueUrl(queueUrl)
            .messageBody(objectMapper.writeValueAsString(message))
            .messageAttributes(buildMessageAttributes())
            .build()
        sqsAsyncClient.sendMessage(request)
    }

    private fun getQueueFromCacheOrSdk(queueName: String): String {
        if (cache.contains(queueName)) {
            logger.debug("Using queueUrl from cache. QueueName: {}", queueName)
            return cache.get(queueName)!!
        }
        return synchronized(cacheLock) {
            if (!cache.contains(queueName)) { // double check to avoid dirty read
                logger.debug("Seaching queueUrl with Amazon SDK. QueueName: {}", queueName)
                cache.add(queueName, searchQueueUrl(queueName))
            }
            cache.get(queueName)!!
        }
    }

    // Do sync because will be done only one time per queue and instance running
    private fun searchQueueUrl(queueName: String): String {
        return sqsAsyncClient.getQueueUrl(GetQueueUrlRequest.builder()
            .queueName(queueName)
            .build()
        ).join().queueUrl()
    }

    private fun buildMessageAttributes(): Map<String, MessageAttributeValue> {
        val attributes: MutableMap<String, MessageAttributeValue> = mutableMapOf()

        openTelemetry.propagators.textMapPropagator.inject(
            Context.current(),
            attributes,
            SqsTracingInjector
        )

        return attributes.toMap()
    }
}

class Cache<K, V> {
    private val cache: MutableMap<K, V> = mutableMapOf()
    fun contains(key: K): Boolean = cache.containsKey(key)
    fun get(key: K): V? = cache[key]
    fun add(key: K, value: V) { cache[key] = value }
}